src/backend/cdb/cdbmutate.c

```diff
@@ -845,7 +845,10 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 	if (IsA(newnode, Motion) &&flow->req_move != MOVEMENT_NONE)
 	{
 		plan = ((Motion *) newnode)->plan.lefttree;
-		flow = plan->flow;
+
+		/* Save previous flow and copy flow of deleting motion. */
```

Comment:
```sql
explain (costs off, verbose) insert into t2 (a, b) select i, random() from t;
```

Reply:
Do you remember what you meant a few months later? Why should we preserve Motion's flow? Why should we save previous flow too?

Reply:
> Why should we preserve Motion's flow? Why should we save previous flow too?

because motion will be deleted and it contains broadcast flow

Reply:
Why is it a problem? Describe it for reviewers. How does it resolve the problem?

Reply:
If we solved the problem with motion removing, still do we need SingleQE transformation before each broadcast instead of explicit motion creation?

Reply:
> If we solved the problem with motion removing, still do we need SingleQE transformation before each broadcast instead of explicit motion creation?

yes, I point at every code change which query breaks without this change

Reply:
I still don't see good explanation for this change.

Reply:
It happens because subquery scan is trrivial in this case
```sql
explain (costs off, verbose) insert into t2 (a, b) select i, random() from t;
             QUERY PLAN              
-------------------------------------
 Insert on public.t2
   ->  Seq Scan on public.t
         Output: t.i, random()
 Optimizer: Postgres query optimizer
(4 rows)
```
while not trivial in case
```sql
explain (costs off, verbose) insert into t2 (a, b) select i, random()::decimal from t;
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Insert on public.t2
   ->  Broadcast Motion 1:3  (slice1; segments: 1)
         Output: "*SELECT*".i, (("*SELECT*".random)::double precision)
         ->  Subquery Scan on "*SELECT*"
               Output: "*SELECT*".i, "*SELECT*".random
               ->  Seq Scan on public.t
                     Output: t.i, (random())::numeric
 Optimizer: Postgres query optimizer
(8 rows)
```